### PR TITLE
[SW-442] Get array size from metadata

### PR DESCRIPTION
**Pull request by petro-rudenko**
Running sparkling water on large datasets (100+gb) noticed it spends a lot of time on figuring out feature vector size: [](https://postimg.org/image/5dggcpw75/). If feature vector was created by Spark's VectorAssembler it would have array size [inside metadata](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala#L88). So what it does - it iterates for all vector/array indicies in metadata and if for all of them there's size in metadata - it can return result in a constant time.

### CODE REVIEW

**Comment by h2o-ops** [comment]
Hi @petro-rudenko, your PR title "Get array size from metadata" is missing JIRA issue number! Please specify it in the following form [SW-]

**Comment by jakubhava** [comment]
Hi @petro-rudenko, thanks for the PR! Let me have a look on it as soon as possible

File: core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala [hunk 1]
```diff
@@ -176,12 +177,27 @@ object H2OSchemaUtils {
     * @return list of max sizes for array types, followed by max sizes for vector types. */
   private[h2o]
   def collectMaxArrays(sc: SparkContext,
-                       rdd: RDD[Row],
+                       dataframe: DataFrame,
                        arrayTypesIndx: Seq[Seq[Int]],
                        vectorTypesIndx: Seq[Seq[Int]]): Array[Int] = {
     val allTypesIndx = arrayTypesIndx ++ vectorTypesIndx
+    val metadataInfo = (field: StructField) => AttributeGroup.fromStructField(field).numAttributes
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala]: I would rename  `metadataInfo` to something like `attributesNum`

File: core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala [hunk 2]
```diff
@@ -176,12 +177,27 @@ object H2OSchemaUtils {
...
+
+    def getBasicType (columns: StructType, indexes: Seq[Int]): StructField = {
+      val structField = columns.fields(indexes.head)
+      if (indexes.tail.isEmpty) structField
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala]: I would use `if(..){
..
}` also here to be consistent, but just my opinion :)

**Comment by petro-rudenko** [comment]
Hey guys, thanks for review. This trick works only with `ml.lilang.VectorUDT` type. But another option is to have some magic key in metadata so user would tell SW the size of arrays type:
`structField.metadata.contains("h2o.array.size")`. 
Recursion is needed to handle nested schema, since `collectVectorLikeTypes` returns `Seq[Seq[Int]]`.

**Comment by jakubhava** [comment]
@petro-rudenko looks good! I will run our internal tests to check if they pass as well. I created JIRA for this issue https://0xdata.atlassian.net/browse/SW-442

In the meanwhile, could you squash to commits and prefix the single commit with the change with `[SW-442]` prefix ? Thanks!!

**Comment by jakubhava** [comment]
All test passed, thanks @petro-rudenko. This change will be in the next release.

**Pull request merged**